### PR TITLE
LIME-153: Upload avatar on profile

### LINKED ISSUE #153: Upload Avatar on Profile Settings Screen
**Avatar photo**

- may be uploaded from the user’s device in PNG and JPEG format
- max size 20 MB
- may be cropped and zoomed manually
- should use Upload File Plugin

### LINKED ISSUE #419: BUG: The blank screen is displayed after clicking on the "Save" button on the Personal Information page
### Describe your issue

The blank screen is displayed instead of loading icon after clicking on the "Save" button on the Personal Information page

### Steps to reproduce

1. Go to the [https://live-in-motion-everyday.com/profile/information](https://live-in-motion-everyday.com/profile/information) as authorized user
2. Click on the "Save" button
3. Pay attention to the page displayed

### What was the actual result?

The blank screen is displayed after clicking on the "Save" button on the Personal Information page

### What was the expected result?

The page with loading icon is displayed after clicking on the "Save" button on the Personal Information page

### Put here any screenshots or videos (optional)

https://github.com/BinaryStudioAcademy/bsa-winter-2023-2024-lime/assets/124812752/f92cdf06-6278-482e-833e-ffadfa3f65cb

### What OS are you seeing the problem on?

Windows

### What browsers are you seeing the problem on?

Chrome

### Severity

Minor (S4)

### LINKED ISSUE #420: Enhancement: Display the loading icon only on the clicked button on the Personal Information page
Display the loading icon only on the clicked button(Save or Reset) on the Personal Information page after trying to update the user's data

Image

### PR DESCRIPTION
**Pull request by JKaypa**

### CODE REVIEW

File: frontend/src/bundles/profile/components/profile-settings/canvas-preview.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
+//@ts-expect-error: unexpected error
```
Comment (nikita-remeslov) [inline frontend/src/bundles/profile/components/profile-settings/canvas-preview.ts]: why do we need it?

File: frontend/src/bundles/common/components/avatar/avatar.tsx [hunk 2]
```diff
@@ -9,7 +9,7 @@ type AvatarSize = Exclude<
 type Properties = {
     size: AvatarSize;
     email: string;
-    avatarUrl: string | null;
+    avatarUrl: string | null | undefined;
```
Comment (nikita-remeslov) [inline frontend/src/bundles/common/components/avatar/avatar.tsx]: why do we need it?

File: frontend/src/bundles/profile/components/profile-settings/cropper.tsx [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+import ReactCrop, {
+    //@ts-expect-error: unexpected error
+    type Crop,
+    //@ts-expect-error: ...
+    type PercentCrop,
+    //@ts-expect-error: ...
+    centerCrop,
+    //@ts-expect-error: ...
+    makeAspectCrop,
+} from 'react-image-crop';
```
Comment (nikita-remeslov) [inline frontend/src/bundles/profile/components/profile-settings/cropper.tsx]: in such way, we can remove TS at all

File: frontend/src/bundles/profile/components/profile-settings/cropper.tsx [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+import { actions as authActions } from '~/bundles/auth/store/auth.js';
+import { Button } from '~/bundles/common/components/components.js';
+import { useAppDispatch } from '~/bundles/common/hooks/hooks.js';
+
+import { canvasPreview } from './canvas-preview.js';
+import { toFile } from './helpers/helpers.js';
+
+const DIMENSION = {
```
Comment (nikita-remeslov) [inline frontend/src/bundles/profile/components/profile-settings/cropper.tsx]: look like it should be constants

File: frontend/src/bundles/profile/components/profile-settings/canvas-preview.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import { type PercentCrop } from 'react-image-crop';
+
+type CanvasProperties = {
```
Comment (nikita-remeslov) [inline frontend/src/bundles/profile/components/profile-settings/canvas-preview.ts]: it should be separate file

File: frontend/.eslintrc.yml [hunk 6]
```diff
@@ -36,3 +36,8 @@ overrides:
       rules:
           unicorn/prevent-abbreviations:
               - off
+    - files:
+          - cropper.tsx
```
Comment (nikita-remeslov) [inline frontend/.eslintrc.yml]: why do we need these changes?

File: backend/src/common/services/file/types/file.ts
Comment (nikita-remeslov) [inline backend/src/common/services/file/types/file.ts:1]: backend/fastify-types.d.ts

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]

File: backend/fastify-types.d.ts [hunk 7]
```diff
@@ -1,9 +1,11 @@
 import { type FastifyRequest as OriginalFastifyRequest } from 'fastify';
+import { type File } from 'fastify-multer/lib/interfaces.js';
```
Comment (dimapopovych) [inline backend/fastify-types.d.ts]: Can we re-export this type somewhere to not import every time from package?

**Review by dimapopovych — CHANGES_REQUESTED** [review]

**Review by nikita-remeslov — COMMENTED** [review]
Sorry but I don't understand why you skip my comments and return ticket to review.
Please check all my comments and provide fixes

**Comment by JKaypa** [comment]
> Sorry but I don't understand why you skip my comments and return ticket to review. Please check all my comments and provide fixes

I'm sorry but they were fixed when you told me, I was just wating for clarification about if I should use JS in the croper file

File: frontend/src/bundles/profile/components/profile-settings/canvas-preview.ts (hunk 1 above)
Comment (nikita-remeslov) [reply frontend/src/bundles/profile/components/profile-settings/canvas-preview.ts]: image

**Review by nikita-remeslov — COMMENTED** [review]
No comments, no fixes
image

No comments, no fixes
image

No comments, no fixes
image

No comments, no fixes
image

What should I do here? Do I need to check this or not? Are you going to fix it or are you going to leave it like this?

The fact that you moved the ticket to review is cool! But the fact that you missed half of the comment and are just waiting is not cool!

How can you move a ticket to review if you haven't fixed all the comments yet? How can I understand what you mean here? If you want some clarification, you should write about it, and not just wait!
Have you written about this somewhere that you need clarification?

Friday report
image
Thursday report:
image
Wednesday report:
image
Tuesday no report
Monday no report

In private messages, also not a single question about this PR!

**Comment by JKaypa** [comment]
I am sorry but here are my comments:
image
image
image
image
Maybe this is the only comment I skip because I thought it was answered here:
image
are these comments not visible?

**Comment by nikita-remeslov** [comment]
image
you no publish your comment it is just a draft and visible only for you

**Comment by stpalamar** [comment]
**Refactored**
@nikita-remeslov


https://github.com/BinaryStudioAcademy/bsa-winter-2023-2024-lime/assets/26843161/5b30f127-e89b-4397-86fe-349bce8b70b3


https://github.com/BinaryStudioAcademy/bsa-winter-2023-2024-lime/assets/26843161/0b855892-43f9-4aa7-b116-c9b9b6b1b2ac

**Comment by stpalamar** [comment]
Updated folder structure

File: frontend/src/bundles/auth/store/slice.ts [hunk 8]
```diff
@@ -6,24 +6,41 @@ import {
...
+    uploadAvatar,
+} from './actions.js';
 
 type State = {
     dataStatus: ValueOf<typeof DataStatus>;
     user: UserAuthResponseDto | null;
     isRefreshing: boolean;
+    updateProfile: {
+        avatarUrl: string | null;
+    };
```
Comment (dimapopovych) [inline frontend/src/bundles/auth/store/slice.ts]: why do we need this additional state?

Comment (stpalamar) [reply frontend/src/bundles/auth/store/slice.ts]: To save avatarUrl after dispatch uploadAvatar.

After fulfilled action I get this new avatarUrl from state and setValue in form and clear this state

**Comment by stpalamar** [comment]
Also, fixed the blank layout while updating user #419 with a temporary solution with separate dataStatus

And made loader only for Save button #420. Loader for **RESET** is redundant because it is not async action

Also I fixed this warnings
image

**Review by nikita-remeslov — DISMISSED** [review]

**Review by dimapopovych — DISMISSED** [review]

**Pull request merged**